tls/policy/s2n_policy_builder.c

```diff
@@ -0,0 +1,187 @@
+    RESULT_GUARD(s2n_kem_preferences_copy(
+            original->kem_preferences, &policy->kem_preferences));
+
+    *copy = policy;
+    ZERO_TO_DISABLE_DEFER_CLEANUP(policy);
+    return S2N_RESULT_OK;
+}
+
+int s2n_security_policy_free(struct s2n_security_policy **policy_ptr)
+{
+    if (!policy_ptr) {
+        return S2N_SUCCESS;
+    }
+    struct s2n_security_policy *policy = *policy_ptr;
+    if (!policy) {
+        return S2N_SUCCESS;
+    }
+
+    /* Static policies should always be const, so this method's non-const
+     * argument should prevent mistakes. However, there are ways to circumvent
+     * const in C, so add an extra safety check.
+     */
+    POSIX_ENSURE(policy->alloced, S2N_ERR_INVALID_ARGUMENT);
+
+    POSIX_GUARD_RESULT(s2n_cipher_preferences_free(&policy->cipher_preferences));
+    POSIX_GUARD_RESULT(s2n_signature_preferences_free(&policy->signature_preferences));
+    POSIX_GUARD_RESULT(s2n_signature_preferences_free(&policy->certificate_signature_preferences));
+    POSIX_GUARD_RESULT(s2n_ecc_preferences_free(&policy->ecc_preferences));
+    POSIX_GUARD_RESULT(s2n_certificate_key_preferences_free(&policy->certificate_key_preferences));
+    POSIX_GUARD_RESULT(s2n_kem_preferences_free(&policy->kem_preferences));
+
+    POSIX_GUARD(s2n_free_object((uint8_t **) policy_ptr, sizeof(struct s2n_security_policy)));
+    return S2N_SUCCESS;
+}
+
+struct s2n_security_policy_builder *s2n_security_policy_builder_from_version(const char *version)
+{
+    PTR_ENSURE(version, S2N_ERR_INVALID_ARGUMENT);
+    DEFER_CLEANUP(struct s2n_blob mem = { 0 }, s2n_free);
+    PTR_GUARD_POSIX(s2n_alloc(&mem, sizeof(struct s2n_security_policy_builder)));
+    struct s2n_security_policy_builder *builder =
+            (struct s2n_security_policy_builder *) (void *) mem.data;
+
+    PTR_GUARD_POSIX(s2n_find_security_policy_from_version(version, &builder->base_policy));
+    PTR_ENSURE_REF(builder->base_policy);
+    PTR_ENSURE(!builder->base_policy->alloced, S2N_ERR_INVALID_SECURITY_POLICY);
+
+    ZERO_TO_DISABLE_DEFER_CLEANUP(mem);
+    return builder;
+}
+
+int s2n_security_policy_builder_free(struct s2n_security_policy_builder **builder_ptr)
+{
+    if (!builder_ptr) {
+        return S2N_SUCCESS;
+    }
+    struct s2n_security_policy_builder *builder = *builder_ptr;
+    *builder_ptr = NULL;
+    if (!builder) {
+        return S2N_SUCCESS;
+    }
+    POSIX_GUARD(s2n_free_object((uint8_t **) &builder, sizeof(struct s2n_security_policy_builder)));
+    return S2N_SUCCESS;
+}
+
+/* For now, "build" just copies the input policy.
+ * We will add functionality later.
+ */
+struct s2n_security_policy *s2n_security_policy_build(struct s2n_security_policy_builder *builder)
+{
+    PTR_ENSURE(builder, S2N_ERR_INVALID_ARGUMENT);
+    struct s2n_security_policy *policy = NULL;
+    PTR_GUARD_RESULT(s2n_security_policy_copy(builder->base_policy, &policy));
+    PTR_ENSURE_REF(policy);
+    policy->alloced = true;
```

Comment:
nit - Is this the best place to set `policy->alloced`? I wonder if it might be best to set it when the policy is actually alloced, in `s2n_security_policy_copy`? Just thinking if we use `s2n_security_policy_copy` somewhere else, we'd have to remember to manually set `alloced` for it there as well.